ccl/changefeedccl: TestChangefeedHandlesDrainingNodes failed

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://mesolite.cluster.engflow.com/invocations/default/1a8c343a-f014-4ccc-98ab-066d4a2eb894?testReportRun=1&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [d65a04b356e5a495fc02c1f37be92b5941913a02](https://github.com/cockroachdb/cockroach/commits/d65a04b356e5a495fc02c1f37be92b5941913a02):

Fatal error:

```
panic: found 1 short-living non-stopped monitors in flow 20ded82d
flow 20ded82d 0 B
    changeagg-mem 0 B


```
Stack: 

```
goroutine 74989 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc00ab62e60, {0x81de220, 0xc008f27b30}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0xc00bd60f08, {0x81de220, 0xc008f27b30})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0xc00f518480, {0x81de220, 0xc008f27b30})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0xfe
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x24
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc007a972c0, {0x81de220, 0xc008f27ae0}, 0xc00403d520, 0x0, 0xc00e5a7b80, 0xc0103efc08, 0xc004351008, 0xc002e25f10)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0xd2c
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x81de220, 0xc008f27a90})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x43d
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x1f
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 74945
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestChangefeedHandlesDrainingNodes2404647006
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39514

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15628777?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15628777?buildTab=artifacts#/) on master @ [f5e65c553608fe6e1c0db6b2a8b279b33790ce2e](https://github.com/cockroachdb/cockroach/commits/f5e65c553608fe6e1c0db6b2a8b279b33790ce2e):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes3991835066
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
panic: found 1 short-living non-stopped monitors in flow e02b85bb
flow e02b85bb 0 B
    changeagg-mem 0 B


goroutine 89039 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x400c9aa5a0, {0x79edb58, 0x4011a09180}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x400b9e3908, {0x79edb58, 0x4011a09180})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x400aa5a360, {0x79edb58, 0x4011a09180})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x40018948c0, {0x79edb58, 0x4011a09130}, 0x4014b5d2b0, 0x0, 0x40151e3400, 0x4012444c08, 0x400bc56808, 0x400b0bbf00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79edb58, 0x4011a090e0})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 89084
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15629494?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15629494?buildTab=artifacts#/) on master @ [d65a04b356e5a495fc02c1f37be92b5941913a02](https://github.com/cockroachdb/cockroach/commits/d65a04b356e5a495fc02c1f37be92b5941913a02):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes554211908
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making pulsar feed factory
panic: found 1 short-living non-stopped monitors in flow b7812884
flow b7812884 0 B
    changeagg-mem 0 B


goroutine 95316 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x40108a6b40, {0x79edcf8, 0x4008476640}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x401491b688, {0x79edcf8, 0x4008476640})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x401af183c0, {0x79edcf8, 0x4008476640})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x40015a6f00, {0x79edcf8, 0x40084765f0}, 0x400e537d40, 0x0, 0x4018e45380, 0x40044c7808, 0x401900d008, 0x4012643f00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79edcf8, 0x4008476550})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 94624
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15630408?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15630408?buildTab=artifacts#/) on master @ [daf384afdcd7b6968414e3d4826503a295bc0590](https://github.com/cockroachdb/cockroach/commits/daf384afdcd7b6968414e3d4826503a295bc0590):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes394968732
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making webhook feed factory
panic: found 1 short-living non-stopped monitors in flow 1f4273f0
flow 1f4273f0 0 B
    changeagg-mem 0 B


goroutine 87957 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x400ffd00a0, {0x79edcf8, 0x40099e3ea0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x400fa19b88, {0x79edcf8, 0x40099e3ea0})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x4008fd1410, {0x79edcf8, 0x40099e3ea0})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x4009516dc0, {0x79edcf8, 0x40099e3e50}, 0x40111d3520, 0x0, 0x4008f72d00, 0x40159e2c08, 0x400ffc4008, 0x40100b5f00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79edcf8, 0x40099e3e00})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 87869
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15632011?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15632011?buildTab=artifacts#/) on master @ [3dfc3dd881da10825b212102913f188a5d472394](https://github.com/cockroachdb/cockroach/commits/3dfc3dd881da10825b212102913f188a5d472394):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes731697604
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
panic: found 1 short-living non-stopped monitors in flow 3cccf422
flow 3cccf422 0 B
    changeagg-mem 0 B


goroutine 103607 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x400a7328c0, {0x79efb18, 0x400879d450}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x40111d6f08, {0x79efb18, 0x400879d450})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x400c5d6360, {0x79efb18, 0x400879d450})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x400d0cdcc0, {0x79efb18, 0x400879d400}, 0x40147161a0, 0x0, 0x400f3a7700, 0x4004588c08, 0x400d0d2808, 0x4002b19f00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79efb18, 0x400879d3b0})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 103492
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15633206?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15633206?buildTab=artifacts#/) on master @ [f414f7c2fa09dc73ce463050afdfe34d6e2246da](https://github.com/cockroachdb/cockroach/commits/f414f7c2fa09dc73ce463050afdfe34d6e2246da):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes1518715630
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
panic: found 1 short-living non-stopped monitors in flow 60e63054
flow 60e63054 0 B
    changeagg-mem 0 B


goroutine 95901 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x40164f3d60, {0x79f0698, 0x4013b45cc0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x40044c6788, {0x79f0698, 0x4013b45cc0})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x400c9795c0, {0x79f0698, 0x4013b45cc0})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x4009e33900, {0x79f0698, 0x4013b45c70}, 0x4018e129c0, 0x0, 0x400c0da980, 0x401100a408, 0x400a9de808, 0x400d677f00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79f0698, 0x4013b45c20})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 96179
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15636270?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15636270?buildTab=artifacts#/) on master @ [8ff30a0d71ab9a20506e2e6940c7bae8dddaeae2](https://github.com/cockroachdb/cockroach/commits/8ff30a0d71ab9a20506e2e6940c7bae8dddaeae2):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes890122991
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making webhook feed factory
panic: found 1 short-living non-stopped monitors in flow 6ce50042
flow 6ce50042 0 B
    changeagg-mem 0 B


goroutine 98241 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x4014570000, {0x79f2658, 0x4013bd5d10}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x401455c508, {0x79f2658, 0x4013bd5d10})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x4014559260, {0x79f2658, 0x4013bd5d10})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x40061c12c0, {0x79f2658, 0x4013bd5cc0}, 0x40139e16c0, 0x0, 0x4014554a80, 0x400c3bd008, 0x401456c008, 0x401470ff00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79f2658, 0x4013bd5c70})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 98103
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://tanzanite.cluster.engflow.com/invocations/default/d30c8125-4a48-445d-b587-0d955986d5c4?testReportRun=2&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4](https://github.com/cockroachdb/cockroach/commits/9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4):

Fatal error:

```
panic: found 1 short-living non-stopped monitors in flow debf0f5e
flow debf0f5e 0 B
    changeagg-mem 0 B


```
Stack: 

```
goroutine 80196 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc009e1b680, {0x81e3300, 0xc00bd28780}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0xc007ab9908, {0x81e3300, 0xc00bd28780})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:718 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0xc010868960, {0x81e3300, 0xc00bd28780})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0xfe
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x24
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc001d25cc0, {0x81e3300, 0xc00bd28730}, 0xc006b0d1e0, 0x0, 0xc00c481780, 0xc00d0d4c08, 0xc007d30008, 0xc00b7d9f10)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0xd2c
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x81e3300, 0xc00bd286e0})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x43d
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x1f
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 80142
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestChangefeedHandlesDrainingNodes1761631016
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15645118?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15645118?buildTab=artifacts#/) on master @ [959a0930e8ce1a9cb839fdb5ac747391dbaf94a3](https://github.com/cockroachdb/cockroach/commits/959a0930e8ce1a9cb839fdb5ac747391dbaf94a3):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/caa72099da386999b346f93d4299a93a/logTestChangefeedHandlesDrainingNodes510724944
    test_log_scope.go:81: use -show-logs to present logs inline
    helpers_test.go:943: making kafka feed factory
panic: found 1 short-living non-stopped monitors in flow b3dd3357
flow b3dd3357 0 B
    changeagg-mem 0 B


goroutine 88600 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0x4009703900, {0x79f26d8, 0x400bb603c0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x57c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x400a510788, {0x79f26d8, 0x400bb603c0})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:717 +0x16c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x4008cb55c0, {0x79f26d8, 0x400bb603c0})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3()
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x4001ae5cc0, {0x79f26d8, 0x400bb60320}, 0x40097196c0, 0x0, 0x400e6be880, 0x40098fe808, 0x401528f008, 0x400df63f00)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:933 +0x9f0
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func1({0x79f26d8, 0x400bb60050})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:327 +0x338
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 88567
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -428,16 +428,12 @@ func (ca *changeAggregator) startKVFeed(
 		defer kvFeedMemMon.Stop(ctx)
 		errCh <- kvfeed.Run(ctx, kvfeedCfg)
 	}); err != nil {
+		// Ensure that the memory monitor is closed properly.
+		kvFeedMemMon.Stop(ctx)
 		return nil, nil, nil, err
 	}
 
 	return buf, doneCh, errCh, nil
-}
-
-func (ca *changeAggregator) waitForKVFeedDone() {
-	if ca.kvFeedDoneCh != nil {
-		<-ca.kvFeedDoneCh
-	}
 }
 
 func (ca *changeAggregator) checkKVFeedErr() error {
@@ -596,14 +592,16 @@ func (ca *changeAggregator) close() {
 	if ca.Closed {
 		return
 	}
-	if ca.cancel == nil {
-		// consumer close may be called even before Start is called.
-		// If that's the case, cancel is not initialized.
-		return
+	if ca.cancel != nil {
+		// consumer close may be called even before Start is called. If that's
+		// the case, cancel is not initialized. We still need to perform the
+		// remainder of the cleanup though.
+		ca.cancel()
 	}
-	ca.cancel()
-	// Wait for the poller to finish shutting down.
-	ca.waitForKVFeedDone()
+	// Wait for the poller to finish shutting down if it was started.
+	if ca.kvFeedDoneCh != nil {
+		<-ca.kvFeedDoneCh
+	}
 
 	if ca.eventConsumer != nil {
 		_ = ca.eventConsumer.Close() // context cancellation expected here.
@@ -625,7 +623,6 @@ func (ca *changeAggregator) close() {
 	}
 
 	ca.memAcc.Close(ca.Ctx())
-
 	ca.MemMonitor.Stop(ca.Ctx())
 	ca.InternalClose()
 }

